src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -59,38 +61,60 @@ BlockchainSQLite::BlockchainSQLite(cryptonote::network_type nettype, fs::path db
   }
 }
 
-
 void BlockchainSQLite::create_schema() {
   LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__);
 
 	db.exec(R"(
-CREATE TABLE batched_payments (
+CREATE TABLE batched_payments_accrued (
+    address VARCHAR NOT NULL,
+    amount BIGINT NOT NULL,
+    PRIMARY KEY(address),
+    CHECK(amount >= 0)
+);
+
+CREATE TABLE batched_payments_paid (
     address VARCHAR NOT NULL,
     amount BIGINT NOT NULL,
-    height_earned BIGINT NOT NULL,
     height_paid BIGINT,
-    PRIMARY KEY(address, height_earned),
-    CHECK(amount > 0)
+    PRIMARY KEY(address, height_paid),
+    CHECK(amount >= 0)
 );
-CREATE VIEW accrued_rewards AS
-    SELECT 
-        address,
-        SUM(amount) as amount,
-        MIN(height_earned) as height
-    FROM batched_payments 
-    WHERE height_paid IS NULL 
-    GROUP BY address;
+
+CREATE INDEX batched_payments_paid_height_idx ON batched_payments_paid (height_paid);
 
 CREATE TABLE batch_db_info (
-    height BIGINT NOT NULL
+    height BIGINT NOT NULL,
+    fire_trigger INT NOT NULL
 );
 
-INSERT INTO batch_db_info (height) VALUES (0);
+INSERT INTO batch_db_info (height, fire_trigger) VALUES (0, 1);
 
 CREATE TRIGGER batch_payments_prune_paid
-AFTER UPDATE ON batch_db_info FOR EACH ROW
+AFTER UPDATE ON batch_db_info
+FOR EACH ROW
+BEGIN
+    UPDATE batch_db_info SET fire_trigger = 0;
+    DELETE FROM batched_payments_paid WHERE height_paid < (NEW.height - 10000);
+    UPDATE batch_db_info SET fire_trigger = 1;
+END;
+
+CREATE TRIGGER batch_payments_delete_empty
+AFTER UPDATE ON batched_payments_accrued
+FOR EACH ROW WHEN NEW.amount = 0 
+BEGIN
+    DELETE FROM batched_payments_accrued WHERE address = NEW.address;
+END;
+
+CREATE TRIGGER make_payment AFTER INSERT ON batched_payments_paid
+FOR EACH ROW WHEN (SELECT fire_trigger from batch_db_info)
+BEGIN
+    UPDATE batched_payments_accrued SET amount = (amount - NEW.amount) WHERE address = NEW.address;
+END;
+
+CREATE TRIGGER delete_payment AFTER DELETE ON batched_payments_paid
+FOR EACH ROW WHEN (SELECT fire_trigger from batch_db_info)
 BEGIN
-    DELETE FROM batched_payments WHERE height_paid < (NEW.height - 10000);
+    UPDATE batched_payments_accrued SET amount = (amount + OLD.amount) WHERE address = OLD.address;
```

Comment:
I'm pretty sure this trigger is the cause of the bug.  Because what's happening in the pop blocks is intending to do this sequence:

- Current balance: 0.005.
- Subtract the block reward of 0.005, balance updates to 0.
- Add the paid amount back in; balance updates to 10.

But we have a trigger on the second step that mucks things up: what we end up doing is:
- Current balance: 0.005
- Subtract block reward of 0.005, balance updates to 0
  - This triggers the delete-0-balance-rows trigger, so we end up with *no* row in the database for the address.
- Delete the `_paid` row to implicitly add the paid amount back in
  - This triggers this UPDATE, but because there is no matching row at all on this address in `_accrued`, it doesn't change anything, and we still have no `_accrued` row (and thus balance is implicitly still 0).

So because of that delete trigger, we end up actually doing:

0.005 -> (no row) -> (no row)

rather than the intended
0.005 -> 0 -> 10

The fix here should be simple enough using the same upsert in this trigger that we use when directly adding an amount elsewhere in the code:
```suggestion
    INSERT INTO batched_payments_accrued (address, amount) VALUES (?, ?)
        ON CONFLICT (address) DO UPDATE SET amount = amount + excluded.amount;
```